nni/algorithms/compression/v2/pytorch/pruning/tools/rl_env/channel_pruning_env.py

```diff
@@ -0,0 +1,585 @@
+
+        self.shared_idx = []
+        '''
+        if self.env_params['model_type'] == 'mobilenetv2':  # TODO: to be tested! Share index for residual connection
+            connected_idx = [4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32]  # to be partitioned
+            last_ch = -1
+            share_group = None
+            for c_idx in connected_idx:
+                if self.prunable_ops[c_idx].in_channels != last_ch:  # new group
+                    last_ch = self.prunable_ops[c_idx].in_channels
+                    if share_group is not None:
+                        self.shared_idx.append(share_group)
+                    share_group = [c_idx]
+                else:  # same group
+                    share_group.append(c_idx)
+            self.shared_idx.append(share_group)
+            _logger.info('=> Conv layers to share channels: %s', self.shared_idx)
+        '''
+        self.min_strategy_dict = copy.deepcopy(self.strategy_dict)
+
+        self.buffer_idx = []
+        for _, v in self.buffer_dict.items():
+            self.buffer_idx += v
+
+        _logger.info('=> Prunable layer idx: %s', self.prunable_idx)
+        _logger.info('=> Buffer layer idx: %s', self.buffer_idx)
+        _logger.info('=> Shared idx: %s', self.shared_idx)
+        _logger.info('=> Initial min strategy dict: %s', self.min_strategy_dict)
+
+        # added for supporting residual connections during pruning
+        self.visited = [False] * len(self.prunable_idx)
+        self.index_buffer = {}
+
+    def _extract_layer_information(self):
+        m_list = list(self.model.modules())
+
+        self.data_saver = []
+        self.layer_info_dict = dict()
+        self.wsize_list = []
+        self.flops_list = []
+
+        # extend the forward fn to record layer info
+        def new_forward(m):
+            def lambda_forward(x):
+                m.input_feat = x.clone()
+                measure_layer_for_pruning(m, x)
+                y = m.old_forward(x)
+                m.output_feat = y.clone()
+                return y
+            return lambda_forward
+
+        device = None
+        for idx in self.prunable_idx + self.buffer_idx:  # get all
+            m = m_list[idx]
+            m.old_forward = m.forward
+            m.forward = new_forward(m)
+            for config in self.config_list:
+                for name in config['op_names']:
+                    if m == get_module_by_name(self.model, name)[-1]:
+                        device = m.weight.device
+
+        # now let the image flow
+        _logger.info('=> Extracting information...')
+        with torch.no_grad():
+            '''
+            for i_b, (inputs, target) in enumerate(self._val_loader):  # use image from train set
+                if i_b == self.n_calibration_batches:
+                    break
+                # self.data_saver.append((inputs.clone(), target.clone()))
+                input_var = torch.autograd.Variable(inputs).to(device)
+            '''
```

Comment:
can be removed?

Reply:
yes, `input_var` now is replaced by `dummy_input`